TypeError: Cannot read property 'results' of null

My Haraka server died and when I looked to see the cause I found the following.

From what I can see, the `mail_from.is_resolvable.js` plugin was trying to access `connection.transaction.results` when `connection.transaction` has been set to `null`.

Perhaps it was set to `null` because the connection was being cleaned up earlier than usual and the `mail_from.is_resolvable.js` plugin still had an async operation running.

	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] running mail hook in mail_from.is_resolvable plugin
	Mar  2 18:06:26 test haraka[31780]: [INFO] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] client [213.121.237.252] half closed connection
	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] running reset_transaction hooks
	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] running disconnect hooks
	Mar  2 18:06:26 test haraka[31780]: [NOTICE] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] disconnect ip=213.121.237.252 rdns="host213-121-237-252.range213-121.btcentralplus.com" helo="AmericanExprecc.com" relay=N early=N esmtp=Y tls=N pipe=N txns=1 rcpts=0/0/0 msgs=0/0/0 bytes=0 lr="" time=235.684
	Mar  2 18:06:26 test haraka[31780]: [CRIT] [-] [core] TypeError: Cannot read property 'results' of null
	Mar  2 18:06:26 test haraka[31780]: [CRIT] [-] [core]     at Plugin.exports.mxErr (/usr/lib/node_modules/Haraka/plugins/mail_from.is_resolvable.js:120:27)
	Mar  2 18:06:26 test haraka[31780]: [CRIT] [-] [core]     at /usr/lib/node_modules/Haraka/plugins/mail_from.is_resolvable.js:54:27
	Mar  2 18:06:26 test haraka[31780]: [CRIT] [-] [core]     at asyncCallback (dns.js:68:16)
	Mar  2 18:06:26 test haraka[31780]: [CRIT] [-] [core]     at Object.onanswer [as oncomplete] (dns.js:149:9)
	Mar  2 18:06:26 test haraka[31780]: [NOTICE] [-] [core] Shutting down

Should plugins never assume `connection.transaction` is present?

Should plugins assume that `connection.transaction` is present when a hook is executed but not after the hook runs an async operation?

Full logs for connection below they are helpful.

	Mar  2 18:02:31 test haraka[31780]: [NOTICE] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] connect ip=213.121.237.252 port=55759 local_ip=:: local_port=25
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running lookup_rdns hooks
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running connect hooks
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running connect hook in access plugin
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [access] checking 213.121.237.252 against connect.rdns_access.whitelist
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [access] checking 213.121.237.252 against connect.rdns_access.whitelist_regex
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [access] checking host213-121-237-252.range213-121.btcentralplus.com against connect.rdns_access.whitelist
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [access] checking host213-121-237-252.range213-121.btcentralplus.com against connect.rdns_access.whitelist_regex
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [access] checking 213.121.237.252 against connect.rdns_access.blacklist_regex
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [access] checking host213-121-237-252.range213-121.btcentralplus.com against connect.rdns_access.blacklist_regex
	Mar  2 18:02:31 test haraka[31780]: [INFO] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [access] pass:unlisted(conn)
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=connect plugin=access function=rdns_access params="" retval=CONT msg=""
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running connect hook in access plugin
	Mar  2 18:02:31 test haraka[31780]: [INFO] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [access] pass:unlisted(conn), connect:any
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=connect plugin=access function=any params="" retval=CONT msg=""
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running connect hook in dnsbl plugin
	Mar  2 18:02:31 test haraka[31780]: [ERROR] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [dnsbl] no zones
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=connect plugin=dnsbl function=connect_first params="" retval=CONT msg=""
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running connect hook in max_unrecognized_commands plugin
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=connect plugin=max_unrecognized_commands function=hook_connect params="" retval=CONT msg=""
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running connect hook in relay plugin
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [relay] checking 213.121.237.252 in relay_acl_allow
	Mar  2 18:02:31 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=connect plugin=relay function=acl params="" retval=CONT msg=""
	Mar  2 18:02:31 test haraka[31780]: [PROTOCOL] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] S: 220 removed.com ESMTP Haraka 2.6.0 ready
	Mar  2 18:02:35 test haraka[31780]: [PROTOCOL] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] C: EHLO AmericanExprecc.com state=1
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running ehlo hooks
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running ehlo hook in access plugin
	Mar  2 18:02:36 test haraka[31780]: [INFO] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [access] pass:unlisted(conn), connect:any, helo:any
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=ehlo plugin=access function=any params="AmericanExprecc.com" retval=CONT msg=""
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running ehlo hook in helo.checks plugin
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=ehlo plugin=helo.checks function=proto_mismatch_esmtp params="AmericanExprecc.com" retval=CONT msg=""
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running ehlo hook in helo.checks plugin
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=ehlo plugin=helo.checks function=init params="AmericanExprecc.com" retval=CONT msg=""
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running ehlo hook in helo.checks plugin
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=ehlo plugin=helo.checks function=match_re params="AmericanExprecc.com" retval=CONT msg=""
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running ehlo hook in helo.checks plugin
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=ehlo plugin=helo.checks function=bare_ip params="AmericanExprecc.com" retval=CONT msg=""
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running ehlo hook in helo.checks plugin
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=ehlo plugin=helo.checks function=dynamic params="AmericanExprecc.com" retval=CONT msg=""
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running ehlo hook in helo.checks plugin
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=ehlo plugin=helo.checks function=big_company params="AmericanExprecc.com" retval=CONT msg=""
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running ehlo hook in helo.checks plugin
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=ehlo plugin=helo.checks function=literal_mismatch params="AmericanExprecc.com" retval=CONT msg=""
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running ehlo hook in helo.checks plugin
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=ehlo plugin=helo.checks function=valid_hostname params="AmericanExprecc.com" retval=CONT msg=""
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running ehlo hook in helo.checks plugin
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=ehlo plugin=helo.checks function=rdns_match params="AmericanExprecc.com" retval=CONT msg=""
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running ehlo hook in helo.checks plugin
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] hook=ehlo plugin=helo.checks function=forward_dns params="AmericanExprecc.com" retval=CONT msg=""
	Mar  2 18:02:36 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] running capabilities hooks
	Mar  2 18:02:36 test haraka[31780]: [PROTOCOL] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] S: 250-removed.com Hello host213-121-237-252.range213-121.btcentralplus.com [213.121.237.252], Haraka is at your service.
	Mar  2 18:02:36 test haraka[31780]: [PROTOCOL] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] S: 250-PIPELINING
	Mar  2 18:02:36 test haraka[31780]: [PROTOCOL] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] S: 250-8BITMIME
	Mar  2 18:02:36 test haraka[31780]: [PROTOCOL] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] S: 250 SIZE 500000
	Mar  2 18:06:26 test haraka[31780]: [PROTOCOL] [5973D3E9-1500-4E1C-98B9-E94A310C925A] [core] C: MAIL FROM:<noreplay@AmericanExprecc.com> SIZE=11114 state=1
	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] running mail hooks
	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] running mail hook in access plugin
	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [access] checking noreplay@americanexprecc.com against mail_from.access.whitelist
	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [access] checking noreplay@americanexprecc.com against mail_from.access.whitelist_regex
	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [access] checking noreplay@americanexprecc.com against mail_from.access.blacklist_regex
	Mar  2 18:06:26 test haraka[31780]: [INFO] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [access] pass:unlisted(mail)
	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] hook=mail plugin=access function=mail_from_access params="<noreplay@americanexprecc.com>" retval=CONT msg=""
	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] running mail hook in access plugin
	Mar  2 18:06:26 test haraka[31780]: [INFO] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [access] pass:unlisted(conn), connect:any, helo:any, mf/rt:any
	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] hook=mail plugin=access function=any params="<noreplay@americanexprecc.com>" retval=CONT msg=""
	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] running mail hook in mail_from.is_resolvable plugin
	Mar  2 18:06:26 test haraka[31780]: [INFO] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] client [213.121.237.252] half closed connection
	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] running reset_transaction hooks
	Mar  2 18:06:26 test haraka[31780]: [DEBUG] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] running disconnect hooks
	Mar  2 18:06:26 test haraka[31780]: [NOTICE] [5973D3E9-1500-4E1C-98B9-E94A310C925A.1] [core] disconnect ip=213.121.237.252 rdns="host213-121-237-252.range213-121.btcentralplus.com" helo="AmericanExprecc.com" relay=N early=N esmtp=Y tls=N pipe=N txns=1 rcpts=0/0/0 msgs=0/0/0 bytes=0 lr="" time=235.684
	Mar  2 18:06:26 test haraka[31780]: [CRIT] [-] [core] TypeError: Cannot read property 'results' of null
	Mar  2 18:06:26 test haraka[31780]: [CRIT] [-] [core]     at Plugin.exports.mxErr (/usr/lib/node_modules/Haraka/plugins/mail_from.is_resolvable.js:120:27)
	Mar  2 18:06:26 test haraka[31780]: [CRIT] [-] [core]     at /usr/lib/node_modules/Haraka/plugins/mail_from.is_resolvable.js:54:27
	Mar  2 18:06:26 test haraka[31780]: [CRIT] [-] [core]     at asyncCallback (dns.js:68:16)
	Mar  2 18:06:26 test haraka[31780]: [CRIT] [-] [core]     at Object.onanswer [as oncomplete] (dns.js:149:9)
	Mar  2 18:06:26 test haraka[31780]: [NOTICE] [-] [core] Shutting down

--- a/plugins/mail_from.is_resolvable.js
+++ b/plugins/mail_from.is_resolvable.js
@@ -24,7 +24,8 @@ exports.load_ini = function () {
 exports.hook_mail = function(next, connection, params) {
     var plugin    = this;
     var mail_from = params[0];
-    var results   = connection.transaction.results;
+    var txn       = connection.transaction;
+    var results   = txn.results;
 
     // Check for MAIL FROM without an @ first - ignore those here
     if (!mail_from.host) {
@@ -39,7 +40,7 @@ exports.hook_mail = function(next, connection, params) {
         // DNS answer didn't return (UDP)
         connection.loginfo(plugin, 'timed out resolving MX for ' + domain);
         called_next++;
-        results.add(plugin, {err: 'timeout(' + domain + ')'});
+        if (txn) results.add(plugin, {err: 'timeout(' + domain + ')'});
         return next(DENYSOFT, 'Temporary resolver error (timeout)');
     }, ((c.timeout) ? c.timeout : 30) * 1000);
 
@@ -51,6 +52,7 @@ exports.hook_mail = function(next, connection, params) {
     };
 
     dns.resolveMx(domain, function(err, addresses) {
+        if (!txn) return;
         if (err && plugin.mxErr(connection, domain, 'MX', err, mxDone)) return;
 
         if (!addresses || !addresses.length) {
@@ -88,6 +90,7 @@ exports.hook_mail = function(next, connection, params) {
             pending_queries++;
             dns.resolve(addr.exchange, function(err, addresses) {
                 pending_queries--;
+                if (!txn) return;
                 if (err) {
                     results.add(plugin, {msg: err.message});
                     connection.logdebug(plugin, domain + ': MX ' +
@@ -117,8 +120,9 @@ exports.hook_mail = function(next, connection, params) {
 
 exports.mxErr = function (connection, domain, type, err, mxDone) {
     var plugin = this;
-    connection.transaction.results.add(plugin,
-            {msg: domain + ':' + type + ':' + err.message});
+    var txn = connection.transaction;
+    if (!txn) return;
+    txn.results.add(plugin, {msg: domain + ':' + type + ':' + err.message});
     connection.logdebug(plugin, domain + ':' + type + ' => ' + err.message);
     switch (err.code) {
         case 'NXDOMAIN':
@@ -137,6 +141,7 @@ exports.implicit_mx = function (connection, domain, mxDone) {
     var plugin = this;
     var txn = connection.transaction;
     dns.resolve(domain, 'A', function(err, addresses) {
+        if (!txn) return;
         if (err && plugin.mxErr(connection, domain, 'A', err, mxDone)) return;
 
         if (!addresses || !addresses.length) {

